synapse/config/oidc_config.py

```diff
@@ -243,12 +250,34 @@ def generate_config_section(self, config_dir_path, server_name, **kwargs):
 # additional checks in the code.
 OIDC_CONFIG_SCHEMA = {"oneOf": [{"type": "null"}, {"type": "object"}]}
 
+# the top-level schema can contain an "oidc_config" and/or an "oidc_providers".
 MAIN_CONFIG_SCHEMA = {
     "type": "object",
-    "properties": {"oidc_config": OIDC_CONFIG_SCHEMA},
+    "properties": {
+        "oidc_config": OIDC_CONFIG_SCHEMA,
+        "oidc_providers": OIDC_PROVIDER_LIST_SCHEMA,
+    },
 }
 
 
+def _parse_oidc_provider_configs(config: JsonDict,) -> Iterable["OidcProviderConfig"]:
+    """extract and parse the OIDC provider configs from the config dict
+
+    Returns a generator which yields the OidcProviderConfig objects
+    """
+    validate_config(MAIN_CONFIG_SCHEMA, config, ())
+
+    for p in config.get("oidc_providers") or []:
+        yield _parse_oidc_config_dict(p)
+
+    # for backwards-compatibility, it is also possible to provide a single "oidc_config"
+    # object with an "enabled: True" property.
+    oidc_config = config.get("oidc_config")
+    if oidc_config and oidc_config.get("enabled", False):
+        validate_config(OIDC_PROVIDER_CONFIG_SCHEMA, oidc_config, ("oidc_config",))
```

Comment:
Isn't the validated via the code above since `MAIN_CONFIG_SCHEMA` references this?

Reply:
`MAIN_CONFIG_SCHEMA` doesn't enforce that `oidc_config` be an `OIDC_PROVIDER_CONFIG_SCHEMA`, merely that it be an object. See the comments at https://github.com/matrix-org/synapse/blob/74346403ccb559f7084bdcb105cd2d76161f978d/synapse/config/oidc_config.py#L244-L250. I guess I should add more comments here.